File: SoftUni-2.0/JavaFundamentals/ExamProblems/ExamPractice/src/org/softuni/RadioactiveBunnies.java
package org.softuni;

import java.awt.Point;
import java.util.*;

// Advanced C# Exam 11 October 2015

public class RadioactiveBunnies {

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        int[] lairSize = Arrays.stream(in.nextLine().split("\\s+"))
                .mapToInt(Integer::parseInt)
                .toArray();
        String[][] lair = new String[lairSize[0]][lairSize[1]];
        for (int i = 0; i < lairSize[0]; i++) {
            lair[i] = in.nextLine().split("");
        }
        List<String> playerMoves = new ArrayList<>();
        Collections.addAll(playerMoves, in.nextLine().split(""));
        Point playerLocation = getPlayerLocation(lair);
        Point lastPlayerLocation = new Point(-1, -1);
        boolean playerWon = false;
        boolean playerDead = false;

        while (!playerWon && !playerDead) {
            switch (playerMoves.get(0)) {
                case "U":
                    playerLocation.x--;
                    if (pointIsOutOfLair(playerLocation, lair)) {
                        playerWon = true;
                        lastPlayerLocation = new Point(playerLocation.x + 1, playerLocation.y);
                    } else if (lair[playerLocation.x][playerLocation.y].equals("B")) {
                        playerDead = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y);
                    } else {
                        lair[playerLocation.x][playerLocation.y] = "P";
                        lair[playerLocation.x + 1][playerLocation.y] = ".";
                    }
                    break;
                case "D":
                    playerLocation.x++;
                    if (pointIsOutOfLair(playerLocation, lair)) {
                        playerWon = true;
                        lastPlayerLocation = new Point(playerLocation.x - 1, playerLocation.y);
                    } else if (lair[playerLocation.x][playerLocation.y].equals("B")) {
                        playerDead = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y);
                    } else {
                        lair[playerLocation.x][playerLocation.y] = "P";
                        lair[playerLocation.x - 1][playerLocation.y] = ".";
                    }
                    break;
                case "L":
                    playerLocation.y--;
                    if (pointIsOutOfLair(playerLocation, lair)) {
                        playerWon = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y + 1);
                    } else if (lair[playerLocation.x][playerLocation.y].equals("B")) {
                        playerDead = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y);
                    } else {
                        lair[playerLocation.x][playerLocation.y] = "P";
                        lair[playerLocation.x][playerLocation.y + 1] = ".";
                    }
                    break;
                case "R":
                    playerLocation.y++;
                    if (pointIsOutOfLair(playerLocation, lair)) {
                        playerWon = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y - 1);
                    } else if (lair[playerLocation.x][playerLocation.y].equals("B")) {
                        playerDead = true;
                        lastPlayerLocation = new Point(playerLocation.x, playerLocation.y);
                    } else {
                        lair[playerLocation.x][playerLocation.y] = "P";
                        lair[playerLocation.x][playerLocation.y - 1] = ".";
                    }
                    break;
            }
            playerMoves.remove(0);

            List<Point> bunnies = new ArrayList<>();
            for (int row = 0; row < lair.length; row++) {
                for (int col = 0; col < lair[row].length; col++) {
                    if (lair[row][col].equals("B")) {
                        bunnies.add(new Point(row, col));
                    }
                }
            }
            for (Point bunnie : bunnies) {
                Point up = new Point(bunnie.x - 1, bunnie.y);
                Point down = new Point(bunnie.x + 1, bunnie.y);
                Point left = new Point(bunnie.x, bunnie.y - 1);
                Point right = new Point(bunnie.x, bunnie.y + 1);
                if (!pointIsOutOfLair(up, lair)) {
                    if (lair[up.x][up.y].equals("P")) {
                        playerDead = true;
                    }
                    lair[up.x][up.y] = "B";
                }
                if (!pointIsOutOfLair(down, lair)) {
                    if (lair[down.x][down.y].equals("P")) {
                        playerDead = true;
                    }
                    lair[down.x][down.y] = "B";
                }
                if (!pointIsOutOfLair(left, lair)) {
                    if (lair[left.x][left.y].equals("P")) {
                        playerDead = true;
                    }
                    lair[left.x][left.y] = "B";
                }
                if (!pointIsOutOfLair(right, lair)) {
                    if (lair[right.x][right.y].equals("P")) {
                        playerDead = true;
                    }
                    lair[right.x][right.y] = "B";
                }
            }
        }
        if (lastPlayerLocation.x == -1 || lastPlayerLocation.y == -1) {
            lastPlayerLocation = playerLocation;
        }
        if (playerWon) {
            if (lair[lastPlayerLocation.x][lastPlayerLocation.y].equals("P")) {
                lair[lastPlayerLocation.x][lastPlayerLocation.y] = ".";
            }
            printField(lair);
            System.out.printf("won: %d %d", lastPlayerLocation.x, lastPlayerLocation.y);
        } else if (playerDead) {
            printField(lair);
            System.out.printf("dead: %d %d", lastPlayerLocation.x, lastPlayerLocation.y);
        }
    }

    private static void printField(String[][] lair) {
        for (String[] row : lair) {
            for (String col : row) {
                System.out.print(col);
            }
            System.out.println();
        }
    }

    private static boolean pointIsOutOfLair(Point location, String[][] lair) {
        return location.x < 0 || location.y < 0 ||
                location.x >= lair.length || location.y >= lair[0].length;
    }

    private static Point getPlayerLocation(String[][] lair) {
        for (int row = 0; row < lair.length; row++) {
            for (int col = 0; col < lair[row].length; col++) {
                if (lair[row][col].equals("P")) {
                    return new Point(row, col);
                }
            }
        }
        return new Point(-1, -1);
    }
}

File: C#-WebDeveloper-3.0/Advanced-C#-May-2016/Exercises/SetsAndDictionaries/RadioactiveBunnies/RadioactiveBunnies.cs
namespace ProblemsWithMatrices
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class Point
    {
        public int X { get; set; }
        public int Y { get; set; }

        public Point(int x, int y)
        {
            this.X = x;
            this.Y = y;
        }

        public Point Move(string direction)
        {
            Point currentPosition = new Point(this.X, this.Y);

            switch (direction)
            {
                case "U":
                    this.X--;
                    break;
                case "D":
                    this.X++;
                    break;
                case "L":
                    this.Y--;
                    break;
                case "R":
                    this.Y++;
                    break;
                default:
                    throw new ArgumentOutOfRangeException();
            }

            return currentPosition;
        }
    }

    public class RadioactiveBunnies
    {
        public static void Main(string[] args)
        {
            var lairSize =
                Console.ReadLine()
                    .Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries)
                    .Select(int.Parse)
                    .ToArray();

            var lair = new string[lairSize[0]][];
            for (int i = 0; i < lairSize[0]; i++)
            {
                lair[i] = Console.ReadLine().ToCharArray().Select(c => c + "").ToArray();
            }

            var playerMoves = Console.ReadLine().ToCharArray().Select(c => c + "").ToList();

            Point playerLocation = GetPlayerLocation(lair);
            Point previousLocation = new Point(playerLocation.X, playerLocation.Y);
            bool playerWon = false;
            bool playerDead = false;

            while (!playerWon && !playerDead)
            {
                previousLocation = playerLocation.Move(playerMoves[0]);

                if (PointIsOutOfLair(playerLocation, lair))
                {
                    playerWon = true;
                }
                else if (lair[playerLocation.X][playerLocation.Y] == "B")
                {
                    playerDead = true;
                }
                else
                {
                    lair[playerLocation.X][playerLocation.Y] = "P";
                    lair[previousLocation.X][previousLocation.Y] = ".";
                }
                playerMoves.RemoveAt(0);

                var bunnies = new List<Point>();
                for (int row = 0; row < lair.Length; row++)
                {
                    for (int col = 0; col < lair[row].Length; col++)
                    {
                        if (lair[row][col] == "B")
                        {
                            bunnies.Add(new Point(row, col));
                        }
                    }
                }

                foreach (Point bunny in bunnies)
                {
                    Point up = new Point(bunny.X - 1, bunny.Y);
                    Point down = new Point(bunny.X + 1, bunny.Y);
                    Point left = new Point(bunny.X, bunny.Y - 1);
                    Point right = new Point(bunny.X, bunny.Y + 1);

                    var directions = new[] { up, down, left, right };

                    foreach (Point direction in directions)
                    {
                        if (!PointIsOutOfLair(direction, lair))
                        {
                            if (lair[direction.X][direction.Y] == "P")
                            {
                                playerDead = true;
                            }
                            lair[direction.X][direction.Y] = "B";
                        }
                    }
                }
            }

            if (playerWon)
            {
                if (lair[previousLocation.X][previousLocation.Y] == "P")
                {
                    lair[previousLocation.X][previousLocation.Y] = ".";
                }
                PrintField(lair);
                Console.WriteLine("won: {0} {1}", previousLocation.X, previousLocation.Y);
            }
            else if (playerDead)
            {
                PrintField(lair);
                Console.WriteLine("dead: {0} {1}", playerLocation.X, playerLocation.Y);
            }
        }

        private static void PrintField(string[][] lair)
        {
            foreach (string[] row in lair)
            {
                foreach (string col in row)
                {
                    Console.Write(col);
                }
                Console.WriteLine();
            }
        }

        private static bool PointIsOutOfLair(Point location, string[][] lair)
        {
            return location.X < 0 || location.Y < 0 || location.X >= lair.Length || location.Y >= lair[0].Length;
        }

        private static Point GetPlayerLocation(string[][] lair)
        {
            for (int row = 0; row < lair.Length; row++)
            {
                for (int col = 0; col < lair[row].Length; col++)
                {
                    if (lair[row][col].Equals("P"))
                    {
                        return new Point(row, col);
                    }
                }
            }
            return new Point(-1, -1);
        }
    }
}
